### PR TITLE
network: introduce StreamInfo in Connection

### PR DESCRIPTION
**Pull request by venilnoronha**
*Description*: This PR removes a reference to `FilterState` from `Connection` in favor of `StreamInfo`.
*Risk Level*: Low
*Testing*: Existing tests suffice
*Docs Changes*: Updated http conn manager doc
*Release Notes*: Added a note about the change
Also see #4748

### CODE REVIEW

File: include/envoy/network/connection.h [hunk 1]
```diff
@@ -240,13 +240,13 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
...
-   * @return the per-connection state associated with this connection.
+   * The StreamInfo object associated with this connection. This is typically
+   * used for logging purposes. Individual filters may add specific information
+   * via the FilterState object within the StreamInfo object.
+   *
+   * @return StreamInfo object associated with this connection.
    */
-  virtual StreamInfo::FilterState& perConnectionState() PURE;
-  virtual const StreamInfo::FilterState& perConnectionState() const PURE;
+  virtual StreamInfo::StreamInfo& streamInfo() PURE;
```
Comment (lizan) [inline include/envoy/network/connection.h]: add back the const interface too. otherwise callers with `const Connection&` cannot read streamInfo.

Comment (venilnoronha) [reply include/envoy/network/connection.h]: Addressed in dc99eb4.

**Comment by rshriram** [comment]
@lizan its not breaking is it? I introduced filter state about 3 weeks ago. Has it made its way into a release already?

File: include/envoy/network/connection.h (hunk 1 above)
Comment (rshriram) [reply include/envoy/network/connection.h]: Why do we need the const interface? Put another way, I did the non const thing by looking at the HTTP filter interface that was supplying a non const StreamInfo.

**Comment by rshriram** [comment]
Hi, as part of this PR, can we also rename the perRequestState to simply filterState() based on the taxonomy doc ? ITs super confusing to do
```
read_callbacks()->connection()->streamInfo()->perRequestState()
``` inside a tcp proxy filter. 

If breaking change is an issue, we can create filterState() functions and mark the perRequestState() as deprecated, route all calls to it to filterState() equivalents.

File: include/envoy/network/connection.h [hunk 2]
```diff
@@ -240,13 +240,14 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
...
   /**
-   * Object on which filters can share data on a per-connection basis.
-   * Only one filter can produce a named data object, but it may be
-   * consumed by many other objects.
-   * @return the per-connection state associated with this connection.
+   * The StreamInfo object associated with this connection. This is typically
+   * used for logging purposes. Individual filters may add specific information
+   * via the FilterState object within the StreamInfo object.
+   *
+   * @return StreamInfo object associated with this connection.
```
Comment (rshriram) [inline include/envoy/network/connection.h]: nit: add a note here that the streamInfo object here is independent of the StreamInfo object inside HTTP connection manager which is per request, while the connection level one is per connection.

File: include/envoy/network/connection.h (hunk 1 above)
Comment (lizan) [reply include/envoy/network/connection.h]: Some filter have const reference/pointer to Connection (e.g. RBAC Engine match), in those case they only need to read StreamInfo hence you need the const interface too. It is not blocking you get a mutable reference from non-const ones.

**Comment by lizan** [comment]
@rshriram yes it is in 1.8 release https://github.com/envoyproxy/envoy/blob/v1.8.0/include/envoy/network/connection.h#L248

I don't think we keep any backward compatibility for C++ APIs, but at least a release note will be helpful.

**Pull request merged**